File: backend/src/main/java/com/nhtl/models/CommandeEcommerce.java
package com.nhtl.models;

import jakarta.persistence.*;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "commandes_ecommerce")
public class CommandeEcommerce {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "user_id", nullable = false)
    private String userId;

    // Infos client
    private String nom;
    private String prenom;
    private String numeroTelephone;
    private String email;

    // Livraison
    private String paysLivraison;
    private String villeLivraison;
    private String adresseLivraison;

    @Enumerated(EnumType.STRING)
    @Column(name = "service_type", nullable = false)
    private ServiceType serviceType;

    @Column(name = "prix_total", nullable = false, precision = 12, scale = 2)
    private BigDecimal prixTotal = BigDecimal.ZERO;

    private String devise = "EUR";

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private EcommerceStatus statut = EcommerceStatus.EN_ATTENTE;

    @Column(nullable = false)
    private Boolean archived = false;

    private String notesSpeciales;

    @Column(name = "date_commande")
    private LocalDateTime dateCommande;

    @Column(name = "date_modification")
    private LocalDateTime dateModification;

    @OneToMany(mappedBy = "commandeEcommerce", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<CommandeEcommerceItem> items = new ArrayList<>();

    @PrePersist
    protected void onCreate() {
        dateCommande     = LocalDateTime.now();
        dateModification = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        dateModification = LocalDateTime.now();
    }

    // ── Getters & Setters ──────────────────────────────────────────────────

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public String getUserId() { return userId; }
    public void setUserId(String userId) { this.userId = userId; }

    public String getNom() { return nom; }
    public void setNom(String nom) { this.nom = nom; }

    public String getPrenom() { return prenom; }
    public void setPrenom(String prenom) { this.prenom = prenom; }

    public String getNumeroTelephone() { return numeroTelephone; }
    public void setNumeroTelephone(String n) { this.numeroTelephone = n; }

    public String getEmail() { return email; }
    public void setEmail(String email) { this.email = email; }

    public String getPaysLivraison() { return paysLivraison; }
    public void setPaysLivraison(String p) { this.paysLivraison = p; }

    public String getVilleLivraison() { return villeLivraison; }
    public void setVilleLivraison(String v) { this.villeLivraison = v; }

    public String getAdresseLivraison() { return adresseLivraison; }
    public void setAdresseLivraison(String a) { this.adresseLivraison = a; }

    public ServiceType getServiceType() { return serviceType; }
    public void setServiceType(ServiceType serviceType) { this.serviceType = serviceType; }

    public BigDecimal getPrixTotal() { return prixTotal; }
    public void setPrixTotal(BigDecimal prixTotal) { this.prixTotal = prixTotal; }

    public String getDevise() { return devise; }
    public void setDevise(String devise) { this.devise = devise; }

    public EcommerceStatus getStatut() { return statut; }
    public void setStatut(EcommerceStatus statut) { this.statut = statut; }

    public Boolean getArchived() { return archived; }
    public void setArchived(Boolean archived) { this.archived = archived; }

    public String getNotesSpeciales() { return notesSpeciales; }
    public void setNotesSpeciales(String n) { this.notesSpeciales = n; }

    public LocalDateTime getDateCommande() { return dateCommande; }
    public void setDateCommande(LocalDateTime dateCommande) { this.dateCommande = dateCommande; }

    public LocalDateTime getDateModification() { return dateModification; }
    public void setDateModification(LocalDateTime dateModification) { this.dateModification = dateModification; }

    public List<CommandeEcommerceItem> getItems() { return items; }
    public void setItems(List<CommandeEcommerceItem> items) { this.items = items; }
}

File: frontend/lib/models/commande_ecommerce.dart
// lib/models/commande_ecommerce.dart

class CommandeEcommerceItem {
  final int? id;
  final int produitId;
  final String? produitNom;
  final int quantite;
  final double prixUnitaire;
  final double sousTotal;
  final String devise;

  CommandeEcommerceItem({
    this.id,
    required this.produitId,
    this.produitNom,
    required this.quantite,
    required this.prixUnitaire,
    required this.sousTotal,
    this.devise = 'EUR',
  });

  factory CommandeEcommerceItem.fromJson(Map<String, dynamic> json) {
    return CommandeEcommerceItem(
      id: (json['id'] as num?)?.toInt(),
      produitId: (json['produitId'] as num?)?.toInt() ?? 0,
      produitNom: json['produitNom'] as String?,
      quantite: (json['quantite'] as num?)?.toInt() ?? 1,
      prixUnitaire: (json['prixUnitaire'] is int)
          ? (json['prixUnitaire'] as int).toDouble()
          : (json['prixUnitaire'] as double? ?? 0.0),
      sousTotal: (json['sousTotal'] is int)
          ? (json['sousTotal'] as int).toDouble()
          : (json['sousTotal'] as double? ?? 0.0),
      devise: json['devise'] as String? ?? 'EUR',
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'produitId': produitId,
        'produitNom': produitNom,
        'quantite': quantite,
        'prixUnitaire': prixUnitaire,
        'sousTotal': sousTotal,
        'devise': devise,
      };
}

class CommandeEcommerce {
  final int? id;
  final String? userId;
  final String nom;
  final String prenom;
  final String numeroTelephone;
  final String? email;
  final String paysLivraison;
  final String villeLivraison;
  final String adresseLivraison;
  final String serviceType;
  final double prixTotal;
  final String devise;
  final String statut;
  final bool archived;
  final String? notesSpeciales;
  final DateTime? dateCommande;
  final DateTime? dateModification;
  final List<CommandeEcommerceItem> items;

  CommandeEcommerce({
    this.id,
    this.userId,
    required this.nom,
    required this.prenom,
    required this.numeroTelephone,
    this.email,
    required this.paysLivraison,
    required this.villeLivraison,
    required this.adresseLivraison,
    required this.serviceType,
    required this.prixTotal,
    this.devise = 'EUR',
    this.statut = 'EN_ATTENTE',
    this.archived = false,
    this.notesSpeciales,
    this.dateCommande,
    this.dateModification,
    this.items = const [],
  });

  factory CommandeEcommerce.fromJson(Map<String, dynamic> json) {
    return CommandeEcommerce(
      id: (json['id'] as num?)?.toInt(),
      userId: json['userId'] as String?,
      nom: json['nom'] as String? ?? '',
      prenom: json['prenom'] as String? ?? '',
      numeroTelephone: json['numeroTelephone'] as String? ?? '',
      email: json['email'] as String?,
      paysLivraison: json['paysLivraison'] as String? ?? '',
      villeLivraison: json['villeLivraison'] as String? ?? '',
      adresseLivraison: json['adresseLivraison'] as String? ?? '',
      serviceType: json['serviceType'] as String? ?? '',
      prixTotal: (json['prixTotal'] is int)
          ? (json['prixTotal'] as int).toDouble()
          : (json['prixTotal'] as double? ?? 0.0),
      devise: json['devise'] as String? ?? 'EUR',
      statut: json['statut'] as String? ?? 'EN_ATTENTE',
      archived: json['archived'] as bool? ?? false,
      notesSpeciales: json['notesSpeciales'] as String?,
      dateCommande: json['dateCommande'] != null
          ? DateTime.tryParse(json['dateCommande'].toString())
          : null,
      dateModification: json['dateModification'] != null
          ? DateTime.tryParse(json['dateModification'].toString())
          : null,
      items: (json['items'] as List<dynamic>?)
              ?.map((i) => CommandeEcommerceItem.fromJson(
                  i as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() => {
        'nom': nom,
        'prenom': prenom,
        'numeroTelephone': numeroTelephone,
        'email': email,
        'paysLivraison': paysLivraison,
        'villeLivraison': villeLivraison,
        'adresseLivraison': adresseLivraison,
        'serviceType': serviceType,
        'devise': devise,
        'notesSpeciales': notesSpeciales,
      };
}
